core/src/main/groovy/noe/tomcat/configure/ConnectorConfiguratorTomcat.groovy

```diff
@@ -124,29 +124,33 @@ class ConnectorConfiguratorTomcat {
   }
 
   private void updateExistingConnector(Node connector, Node newConnector) {
-    // update attributes
+    // update connector attributes
     newConnector.attributes().each { attribute ->
       connector.@"${attribute.key}" = attribute.value
     }
 
-    // update nodes
     newConnector.each { Node newSubelement ->
-      String UpgradeProtocol = ConnectorUpgradeProtocolTomcat.ELEMENT_NAME
-
-      if (newSubelement.name() == UpgradeProtocol) {
-        if (connector.find { it.name() == UpgradeProtocol } == null) {
-          // create new element
-          connector.appendNode(newSubelement, newSubelement.attributes(), newSubelement.value())
-        } else {
-          // upgrade existing element
-          connector.findAll { it.name() == UpgradeProtocol }.each { upgradeProtocol ->
-            newSubelement.attributes() { attribute ->
-              upgradeProtocol.@"${attribute.key}" = attribute.value
-            }
-          }
-        }
+      updateInnerElements(connector, newSubelement)
+    }
+  }
+
+  /**
+   *
+   * @param connector
+   * @param newSubElement
+   *
+   * Search for existing inner element to remove and replace with new one
+   * Note: SSLHostConfig inner elements can exist more than one
+   */
+  private void updateInnerElements(Node connector, Node newSubElement) {
+    if (connector.find { it.name() == newSubElement.name() } != null) {
```

Comment:
What about already existing elements? The methods are called '*update*' ?

Reply:
I noticed cannot be done proper update to inner elements of Connector and is not needed. If someone for example wants to update UpgradeProtocol or SSLHostConfig will write the element with the desired attributes. Then the above function will remove the specific inner element and replace it with the new desired.

Reply:
Let me explain it

Reply:
With conf like: 

`new TomcatConfigurator(tomcat)`
  `.httpsConnector(new SecureHttpConnectorTomcat()`
  `.setPort(8443)`
  `.setUpgradeProtocolToHttp2Protocol()`

Now will add/update the connector with SSLHostConfig element leaving UpgradeProtocol as is:
`Object map = new ConnectorSSLHostConfigTomcat().setCaCertificateFile("server.key")`

`new TomcatConfigurator(tomcat)`
  `.httpsConnector(new SecureHttpConnectorTomcat()`
  `.setSSLHostConfig(map))`

the result will be :

`<Connector port="8080" secure="true" SSLEnabled="true" clientAuth="false" sslEnabledProtocols="false">`
`      <UpgradeProtocol className="org.apache.coyote.http2.Http2Protocol"/>`
`      <SSLHostConfig caCertificateFile="server.key"/>`

Keep in mind that updateInnerElements function looking inside newSubElement node which is the new upcoming node.Matches the inner element with the same name.If it's present, removing the old and create a new with the newSubElement attributes.
You were doing the same pattern matching on the old function.But updating attributes independently wasnt working too good.I just made it to remove whole element and replace it with the fresh

Reply:
I see your point. What I had in a mind is a use-case that on input is completely non-default server.xml and I was wondering whether is OK to change the whole element in case that only for instance 1 attribute is configured only. 

Seems to me, that it should be OK since the 'updateExistingConnector' is part of private API and the client see only `TomcatConfigurator#httpConnector` or so.

I would recommend documenting this behavior into public API though.

Reply:
> I would recommend documenting this behavior into public API though.

That the server.xml will contain "exactly" what is set in code, and what are the implication.

Reply:
Renamed function to replaceInnerElements.Good point Jan !